diff --git a/kernel/riscv64/axpy_vector.c b/kernel/riscv64/axpy_vector.c
index c77a18afa..041f797a6 100644
--- a/kernel/riscv64/axpy_vector.c
+++ b/kernel/riscv64/axpy_vector.c
@@ -29,7 +29,7 @@ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "common.h"
 
 #ifdef RISCV64_ZVL256B
-#       define LMUL m2
+#       define LMUL m4
 #       if defined(DOUBLE)
 #               define ELEN 64
 #       else
