diff --git a/Configure b/Configure
index 75f1f6b67b..4898c43a3a 100755
--- a/Configure
+++ b/Configure
@@ -635,7 +635,6 @@ our %disabled = ( # "what"         => "comment"
                   "h3demo"              => "default",
                   "hqinterop"           => "default",
                   "ec_explicit_curves" => "default",
-                  "ec_nistp_64_gcc_128" => "default",
                   "egd"                 => "default",
                   "engine"              => "default",
                   "external-tests"      => "default",
