diff --git a/modules/video/src/dis_flow.cpp b/modules/video/src/dis_flow.cpp
index c859555c22..a35fcd7126 100644
--- a/modules/video/src/dis_flow.cpp
+++ b/modules/video/src/dis_flow.cpp
@@ -498,16 +498,16 @@ DISOpticalFlowImpl::PatchInverseSearch_ParBody::PatchInverseSearch_ParBody(DISOp
 
 /* Some auxiliary macros */
 #define HAL_INIT_BILINEAR_8x8_PATCH_EXTRACTION                                                                         \
-    v_float32x4 w00v = v_setall_f32(w00);                                                                              \
-    v_float32x4 w01v = v_setall_f32(w01);                                                                              \
-    v_float32x4 w10v = v_setall_f32(w10);                                                                              \
-    v_float32x4 w11v = v_setall_f32(w11);                                                                              \
+    v_float32 w00v = v_setall_f32(w00);                                                                              \
+    v_float32 w01v = v_setall_f32(w01);                                                                              \
+    v_float32 w10v = v_setall_f32(w10);                                                                              \
+    v_float32 w11v = v_setall_f32(w11);                                                                              \
                                                                                                                        \
-    v_uint16x8 I0_row_8, I1_row_8, I1_row_shifted_8, I1_row_next_8, I1_row_next_shifted_8, tmp;                        \
-    v_uint32x4 I0_row_4_left, I1_row_4_left, I1_row_shifted_4_left, I1_row_next_4_left, I1_row_next_shifted_4_left;    \
-    v_uint32x4 I0_row_4_right, I1_row_4_right, I1_row_shifted_4_right, I1_row_next_4_right,                            \
+    v_uint16 I0_row_8, I1_row_8, I1_row_shifted_8, I1_row_next_8, I1_row_next_shifted_8, tmp;                        \
+    v_uint32 I0_row_4_left, I1_row_4_left, I1_row_shifted_4_left, I1_row_next_4_left, I1_row_next_shifted_4_left;    \
+    v_uint32 I0_row_4_right, I1_row_4_right, I1_row_shifted_4_right, I1_row_next_4_right,                            \
       I1_row_next_shifted_4_right;                                                                                     \
-    v_float32x4 I_diff_left, I_diff_right;                                                                             \
+    v_float32 I_diff_left, I_diff_right;                                                                             \
                                                                                                                        \
     /* Preload and expand the first row of I1: */                                                                      \
     I1_row_8 = v_load_expand(I1_ptr);                                                                                  \
@@ -560,16 +560,16 @@ inline float processPatch(float &dst_dUx, float &dst_dUy, uchar *I0_ptr, uchar *
                           int I0_stride, int I1_stride, float w00, float w01, float w10, float w11, int patch_sz)
 {
     float SSD = 0.0f;
-#if CV_SIMD128
+#if (CV_SIMD128 || CV_SIMD_SCALABLE)
     if (patch_sz == 8)
     {
         /* Variables to accumulate the sums */
-        v_float32x4 Ux_vec = v_setall_f32(0);
-        v_float32x4 Uy_vec = v_setall_f32(0);
-        v_float32x4 SSD_vec = v_setall_f32(0);
+        v_float32 Ux_vec = v_setall_f32(0);
+        v_float32 Uy_vec = v_setall_f32(0);
+        v_float32 SSD_vec = v_setall_f32(0);
 
-        v_int16x8 I0x_row, I0y_row;
-        v_int32x4 I0x_row_4_left, I0x_row_4_right, I0y_row_4_left, I0y_row_4_right;
+        v_int16 I0x_row, I0y_row;
+        v_int32 I0x_row_4_left, I0x_row_4_right, I0y_row_4_left, I0y_row_4_right;
 
         HAL_INIT_BILINEAR_8x8_PATCH_EXTRACTION;
         for (int row = 0; row < 8; row++)
@@ -627,17 +627,17 @@ inline float processPatchMeanNorm(float &dst_dUx, float &dst_dUy, uchar *I0_ptr,
     float sum_I0x_mul = 0.0, sum_I0y_mul = 0.0;
     float n = (float)patch_sz * patch_sz;
 
-#if CV_SIMD128
+#if (CV_SIMD128 || CV_SIMD_SCALABLE)
     if (patch_sz == 8)
     {
         /* Variables to accumulate the sums */
-        v_float32x4 sum_I0x_mul_vec = v_setall_f32(0);
-        v_float32x4 sum_I0y_mul_vec = v_setall_f32(0);
-        v_float32x4 sum_diff_vec = v_setall_f32(0);
-        v_float32x4 sum_diff_sq_vec = v_setall_f32(0);
+        v_float32 sum_I0x_mul_vec = v_setall_f32(0);
+        v_float32 sum_I0y_mul_vec = v_setall_f32(0);
+        v_float32 sum_diff_vec = v_setall_f32(0);
+        v_float32 sum_diff_sq_vec = v_setall_f32(0);
 
-        v_int16x8 I0x_row, I0y_row;
-        v_int32x4 I0x_row_4_left, I0x_row_4_right, I0y_row_4_left, I0y_row_4_right;
+        v_int16 I0x_row, I0y_row;
+        v_int32 I0x_row_4_left, I0x_row_4_right, I0y_row_4_left, I0y_row_4_right;
 
         HAL_INIT_BILINEAR_8x8_PATCH_EXTRACTION;
         for (int row = 0; row < 8; row++)
@@ -693,10 +693,10 @@ inline float computeSSD(uchar *I0_ptr, uchar *I1_ptr, int I0_stride, int I1_stri
                         float w11, int patch_sz)
 {
     float SSD = 0.0f;
-#if CV_SIMD128
+#if (CV_SIMD128 || CV_SIMD_SCALABLE)
     if (patch_sz == 8)
     {
-        v_float32x4 SSD_vec = v_setall_f32(0);
+        v_float32 SSD_vec = v_setall_f32(0);
         HAL_INIT_BILINEAR_8x8_PATCH_EXTRACTION;
         for (int row = 0; row < 8; row++)
         {
@@ -728,11 +728,11 @@ inline float computeSSDMeanNorm(uchar *I0_ptr, uchar *I1_ptr, int I0_stride, int
 {
     float sum_diff = 0.0f, sum_diff_sq = 0.0f;
     float n = (float)patch_sz * patch_sz;
-#if CV_SIMD128
+#if (CV_SIMD128 || CV_SIMD_SCALABLE)
     if (patch_sz == 8)
     {
-        v_float32x4 sum_diff_vec = v_setall_f32(0);
-        v_float32x4 sum_diff_sq_vec = v_setall_f32(0);
+        v_float32 sum_diff_vec = v_setall_f32(0);
+        v_float32 sum_diff_sq_vec = v_setall_f32(0);
         HAL_INIT_BILINEAR_8x8_PATCH_EXTRACTION;
         for (int row = 0; row < 8; row++)
         {
@@ -758,7 +758,7 @@ inline float computeSSDMeanNorm(uchar *I0_ptr, uchar *I1_ptr, int I0_stride, int
                 sum_diff += diff;
                 sum_diff_sq += diff * diff;
             }
-#if CV_SIMD128
+#if (CV_SIMD128 || CV_SIMD_SCALABLE)
     }
 #endif
     return sum_diff_sq - sum_diff * sum_diff / n;
